lib/src/client.dart

```diff
@@ -396,31 +399,40 @@ class ClientMethod<Q, R> {
   ClientMethod(this.path, this.requestSerializer, this.responseDeserializer);
 }
 
+/// Provides per-RPC metadata.
+///
+/// Metadata providers will be invoked for every RPC, and can add their own
+/// metadata to the RPC. If the function returns a [Future], the RPC will await
+/// completion of the returned [Future] before transmitting the request.
+typedef FutureOr<Null> MetadataProvider(Map<String, String> metadata);
```

Comment:
Please document that the `MetadataProvider` is supposed to mutate the provided `metadata` map before it returns or before its returned `Future` completes. And at the risk of stating the obvious: a `MetadataProvider` may assume that the provided map is not null.